cdbot/cogs/quoting.py

```diff
@@ -0,0 +1,382 @@
+            return Embed.from_dict(quote["embed"])
+        channel = self.bot.get_channel(quote["channel"])
+        quoted = await self.get_member_by_id(quote["author"]["id"])
+        quoter = await self.get_member_by_id(quote["quoted_by"])
+        embed = Embed()
+        if quoted:
+            embed = embed.set_author(name=str(quoted), icon_url=str(quoted.avatar_url_as(size=32)))
+        else:
+            embed = embed.set_author(name=quote["author"]["name"], icon_url=quote["author"]["avatar"])
+        footer_text = "Quoted by "
+        if quoter:
+            footer_text += str(quoter)
+        else:
+            footer_text += f"ID: {quote['quoted_by']}"
+        footer_text += " in channel "
+        if channel:
+            footer_text += "#" + str(channel)
+        else:
+            footer_text += f"ID: {quote['channel']}"
+        if quoter:
+            embed = embed.set_footer(text=footer_text, icon_url=str(quoter.avatar_url_as(size=32)))
+        else:
+            embed = embed.set_footer(text=footer_text)
+        if "```" in quote["content"]:
+            embed.description = quote["content"] + "\n"
+        elif len(quote["content"]) > 0:
+            embed.description = f"```{quote['content']}```\n"
+        else:
+            embed.description = ""
+        embed.description += f"[Jump to orignal]({quote['content_link']})"
+        embed.timestamp = quote["quoted_at"]
+        if quote.get("image", None):
+            embed.set_image(url=quote["image"])
+        return embed
+
+    async def range_quote(self, channel: TextChannel, after: Message, before: Message) -> List[Message]:
+        messages = await channel.history(after=after, before=before).flatten()
+        return [after, *messages, before]
+
+    def multi_quote_dict(self, messages: List[Message], message: Message) -> dict:
+        multi_quote = dict(
+            content_id=messages[0].id,
+            content_link=messages[0].jump_url,
+            channel=messages[0].channel.id,
+            quoted_at=message.created_at,
+            quoted_by=message.author.id,
+            messages=[],
+        )
+        for line in messages:
+            data = dict(original=line.id, content=line.content, author=dict(id=line.author.id, name=str(line.author)))
+            if len(line.embeds) > 0:
+                data["embed"] = line.embeds[0].to_dict()
+            multi_quote["messages"].append(data)
+        return multi_quote
+
+    async def multi_quote_embed(self, multi_quote: dict) -> Embed:
+        embeds = [message.get("embed", None) for message in multi_quote["messages"]]
+        if any(embeds):
+            return list(filter(lambda x: x is not None, embeds))
+        embed = Embed().set_author(name=str(self.bot.user), icon_url=self.bot.user.avatar_url)
+        quoter = await self.get_member_by_id(multi_quote["quoted_by"])
+        channel = self.bot.get_channel(multi_quote["channel"])
+        footer_text = "Quoted by "
```

Comment:
This functionality is the same as L101 onwards (well, onwards a bit) it seems? Can it be moved to a function?

Reply:
A lot of it is similar, but due to the differences in how multiquotes and normal quotes are stored, it'd be a bit more difficult.